phase/runhooks.go

```diff
@@ -0,0 +1,86 @@
+package phase
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/k0sproject/k0sctl/config"
+	"github.com/k0sproject/k0sctl/config/cluster"
+)
+
+var _ phase = &RunHooks{}
+
+// RunHooks phase runs a set of hooks configured for the host
+type RunHooks struct {
+	Action string
+	Stage  string
+
+	steps map[*cluster.Host][]string
+}
+
+// Title for the phase
+func (p *RunHooks) Title() string {
+	return fmt.Sprintf("Run %s %s Hooks", strings.Title(p.Stage), strings.Title(p.Action))
+}
+
+// Prepare digs out the hosts with steps from the config
+func (p *RunHooks) Prepare(config *config.Cluster) error {
+	p.steps = make(map[*cluster.Host][]string)
+	for _, h := range config.Spec.Hosts {
+		if len(h.Hooks) > 0 {
+			p.steps[h] = h.Hooks.ForActionAndStage(p.Action, p.Stage)
+		}
+	}
+
+	return nil
+}
+
+// ShouldRun is true when there are hosts that need to be connected
+func (p *RunHooks) ShouldRun() bool {
+	return len(p.steps) > 0
+}
+
+// Run does all the prep work on the hosts in parallel
+func (p *RunHooks) Run() error {
+	var wg sync.WaitGroup
+	var errors []string
+	type erritem struct {
+		host string
+		err  error
+	}
+	ec := make(chan erritem, 1)
+
+	wg.Add(len(p.steps))
+
+	for h, steps := range p.steps {
+		go func(h *cluster.Host, steps []string) {
+			for _, s := range steps {
+				err := h.Exec(s)
+				if err != nil {
+					ec <- erritem{h.String(), err}
+					return // do not exec remaining steps if one fails
+				}
+			}
+
+			ec <- erritem{h.String(), nil}
+		}(h, steps)
+	}
+
+	go func() {
```

Comment:
hmmmm. there's no `wg.Add` for the error reader thing, but it has a `wg.Done`. I think there should be a close `ec` after `wg.Wait()`  and the error chan reader would finish because the channel gets closed 🤔

Reply:
hmm no, actually it does exactly what it is supposed to, it's just weird.

Reply:
the err reader does the wg.dones "for" the step-runner gofunc, each of them send to the err channel on finish.

Reply:
I think I need to refactor this anyway to use the `hosts.ParallelEach` helper

Reply:
Yeah, I think the `ParallelEach` functionality was rewritten here in attempt to make the hooks-runner phase work "product agnostically" in LP.

Reply:
Did the refactoring